### PR TITLE
Loader: Implement NCCH encryption check

### PR DESCRIPTION
**Pull request by Gareth422**
I've implemented a simple encryption check which compares the program IDs in the NCCH header and the extended header to see if the ROM is encrypted. I changed all of the header values to be explicitly little-endian to prevent any surprises if Citra gets ported to a big-endian microwave oven. I also changed the program ID from a uint8 array to a u64_le.

Issue #708

### CODE REVIEW

**Comment by yuriks** [comment]
Micronit: Write commit messages in the imperative ("implement" instead of "implemented"), also, refer to the module you're modifying: "Loader: Implement NCCH encryption check".

File: src/core/loader/loader.h [hunk 1]
```diff
@@ -14,7 +14,7 @@
 #include "common/file_util.h"
 
 namespace Kernel {
-struct AddressMapping;
+    struct AddressMapping;
```
Comment (yuriks) [inline src/core/loader/loader.h]: We don't indent namespace contents. Also, spurious unrelated change.

File: src/core/loader/ncch.cpp [hunk 2]
```diff
@@ -128,7 +128,7 @@ ResultStatus AppLoader_NCCH::LoadExec() {
     if (ResultStatus::Success == ReadCode(code)) {
         std::string process_name = Common::StringFromFixedZeroTerminatedBuffer(
                 (const char*)exheader_header.codeset_info.name, 8);
-        u64 program_id = *reinterpret_cast<u64_le const*>(&ncch_header.program_id[0]);
+        u64 program_id = ncch_header.program_id;
```
Comment (yuriks) [inline src/core/loader/ncch.cpp]: I don't think this variable is necessary anymore, since now you can just read the program_id field directly. Just substitute it's uses below.

File: src/core/loader/loader.h (hunk 1 above)
Comment (lioncash) [inline src/core/loader/loader.h]: this isn't related to the change

File: src/core/loader/ncch.cpp [hunk 3]
```diff
@@ -278,6 +283,7 @@ ResultStatus AppLoader_NCCH::Load() {
     if (file.ReadBytes(&exefs_header, sizeof(ExeFs_Header)) != sizeof(ExeFs_Header))
         return ResultStatus::Error;
 
+
```
Comment (lioncash) [inline src/core/loader/ncch.cpp]: unnecessary newline

File: src/core/loader/ncch.cpp [hunk 4]
```diff
@@ -266,6 +266,11 @@ ResultStatus AppLoader_NCCH::Load() {
     LOG_DEBUG(Loader, "Thread priority:             0x%X"  , priority);
     LOG_DEBUG(Loader, "Resource limit category:     %d"    , resource_limit_category);
 
+    if (exheader_header.arm11_system_local_caps.program_id != ncch_header.program_id) {
+        LOG_ERROR(Loader, "Program ID mismatch: the ROM is probably encrypted.");
```
Comment (yuriks) [inline src/core/loader/ncch.cpp]: "ExHeader program id mismatch: [...]"

File: src/core/loader/ncch.cpp (hunk 3 above)
Comment (yuriks) [inline src/core/loader/ncch.cpp]: Unnecessary extra newline

File: src/core/loader/ncch.h [hunk 5]
```diff
@@ -17,31 +17,31 @@
 
 struct NCCH_Header {
     u8 signature[0x100];
-    u32 magic;
-    u32 content_size;
+    u32_le magic;
```
Comment (yuriks) [inline src/core/loader/ncch.h]: Split these changes into a separate commit.

File: src/core/loader/ncch.h [hunk 6]
```diff
@@ -109,7 +109,7 @@ struct ExHeader_StorageInfo {
 };
 
 struct ExHeader_ARM11_SystemLocalCaps {
-    u8 program_id[8];
+    u64 program_id;
```
Comment (yuriks) [inline src/core/loader/ncch.h]: This should also be u64_le.

**Comment by Gareth422** [comment]
I've reorganised all of my changes into two commits.

**Comment by mailwl** [comment]
i like it

**Comment by yuriks** [comment]
The commit messages need to have the "Loader: " or "NCCH:" prefix to indicate what they're modifying. Other than that, LGTM

**Comment by Gareth422** [comment]
I've changed the commit messages.

**Comment by Road-Drum** [comment]
good idea. helps to reduce "failed to load rom" issues (mainly because encrypted rom)

**Comment by Gareth422** [comment]
> good idea. helps to reduce "failed to load rom" issues (mainly because encrypted rom)

That's the idea.

File: src/core/loader/ncch.cpp [hunk 7]
```diff
@@ -266,6 +265,11 @@ ResultStatus AppLoader_NCCH::Load() {
     LOG_DEBUG(Loader, "Thread priority:             0x%X"  , priority);
     LOG_DEBUG(Loader, "Resource limit category:     %d"    , resource_limit_category);
 
+    if (exheader_header.arm11_system_local_caps.program_id != ncch_header.program_id) {
+        LOG_ERROR(Loader, "ExHeader Program ID mismatch: the ROM is probably encrypted.");
```
Comment (jroweboy) [inline src/core/loader/ncch.cpp]: Maybe you should include a link to the wiki article on decryption or maybe say something like "Please do not create an issue on Github if you have not decrypted this game. You can decrypt your game by following these instructions: [wiki link]" My wording is poor, but saying something to that effect here might help a bit with the issue spam (I know they probably won't know how to copy paste the url, but its better than not having it IMO)

Comment (Gareth422) [inline src/core/loader/ncch.cpp]: Having that in the log might not be that useful, as I'm not sure how many people would read it. I think it would make more sense to put that in the GUI pop-up when it gets implemented.

Comment (jroweboy) [inline src/core/loader/ncch.cpp]: I agree with your sentiment, but I don't see how it could hurt to have it here until a GUI pop up is ready. Its not log spam that developers ever encounter, so, to me at least, it can only have a positive effect in the slim chance a user reads the message and doesn't make a issue about it. (Just my opinion though)

Comment (bunnei) [inline src/core/loader/ncch.cpp]: I think the wording as-is is fine. I agree with the sentiment, but it really belongs in a pop up dialog and not in the debug log; and because we don't really have a nice cross-platform/cross-frontend way of doing this from core code, it's out of scope for this PR I think.

**Comment by bunnei** [comment]
LGTM
